internal/cf/binding.go

```diff
@@ -18,21 +18,41 @@ import (
 	"github.com/sap/cf-service-operator/internal/facade"
 )
 
-func (c *spaceClient) GetBinding(ctx context.Context, owner string) (*facade.Binding, error) {
+// GetBinding returns the binding with the given mapBindingOpts["owner"] or mapBindingOpts["name"].
+// If mapBindingOpts["name"] is empty, the binding with the given mapBindingOpts["owner"] is returned.
+// If mapBindingOpts["name"] is not empty, the binding with the given Name is returned for orphan bindings.
+// If no binding is found, nil is returned.
+// If multiple bindings are found, an error is returned.
+// The function add the parameter values to the orphan cf instance, so that can be adopted.
+func (c *spaceClient) GetBinding(ctx context.Context, mapBindingOpts map[string]string) (*facade.Binding, error) {
 	listOpts := cfclient.NewServiceCredentialBindingListOptions()
-	listOpts.LabelSelector.EqualTo(labelPrefix + "/" + labelKeyOwner + "=" + owner)
+	if mapBindingOpts["name"] != "" {
+		listOpts.Names.EqualTo(mapBindingOpts["name"])
+	} else {
+		listOpts.LabelSelector.EqualTo(fmt.Sprintf("%s/%s=%s", labelPrefix, labelKeyOwner, mapBindingOpts["owner"]))
+	}
+
```

Comment:
How about replacing this with
```go
func (c *spaceClient) GetBinding(ctx context.Context, listOpts ServiceCredentialBindingListOptions) (*facade.Binding, error) {
```

And then just creating the listOpts at the servicebinding_controller level?

Reply:
**The suggested change is not viable.** ServiceCredentialBindingListOptions is part of the go-cfclient library's client package. 

The package controller does not have direct access to the go-cfclient library's client package. Instead, it has an interface called SpaceClient, which wraps any interaction with the go-client library.

That is a problem because changing the function GetBinding to receive listOpts will break the designed separation between the controller and the go-cfclient library.

Also, we will have to create the LabelSelector in the controller:
listOpts.LabelSelector.EqualTo(fmt.Sprintf("%s/%s=%s", labelPrefix, labelKeyOwner, mapBindingOpts["owner"]))

labelPrefix: is only visible within the cf package, not the controller package.
labelKeyOwner: is only visible within the cf package, not the controller package.

Again, this breaks the current design where those variables are only visible within the cf package.

Change this:
const (
	labelPrefix                = "service-operator.cf.cs.sap.com"
	labelKeyOwner              = "owner"
	labelOwner                 = labelPrefix + "/" + labelKeyOwner
	annotationPrefix           = "service-operator.cf.cs.sap.com"
	annotationKeyGeneration    = "generation"
	annotationGeneration       = annotationPrefix + "/" + annotationKeyGeneration
	annotationKeyParameterHash = "parameter-hash"
	annotationParameterHash    = annotationPrefix + "/" + annotationKeyParameterHash
)

to 

const (
	**LabelPrefix**                = "service-operator.cf.cs.sap.com"
	**LabelKeyOwner**              = "owner"
	**LabelOwner**                 = labelPrefix + "/" + labelKeyOwner
	**AnnotationPrefix**           = "service-operator.cf.cs.sap.com"
	**AnnotationKeyGeneration**    = "generation"
	**AnnotationGeneration**       = annotationPrefix + "/" + annotationKeyGeneration
	**AnnotationKeyParameterHash** = "parameter-hash"
	**AnnotationParameterHash**    = annotationPrefix + "/" + annotationKeyParameterHash
)

This change is not going to be accepted because we will have to break the separation of concerns between the controller and the cf client.

Reply:
You're right!

I thought about including the gocfclient in the controller might not be very nice, I guess it is worse than I thought since you have to also expose the constants.

Maybe add a new type instead, to pass the owner or name? Like this:

```go
type BindingFilter interface {
	GetListOptions() *cfclient.ServiceCredentialBindingListOptions
}

type BindingFilterName struct {
	name string
}
type BindingFilterOwner struct {
	owner string
}

func (bn *BindingFilterName) GetListOptions() *cfclient.ServiceCredentialBindingListOptions {
	listOpts := cfclient.NewServiceCredentialBindingListOptions()
	listOpts.Names.EqualTo(bn.name)
	return listOpts
}

func (bo *BindingFilterOwner) GetListOptions() *cfclient.ServiceCredentialBindingListOptions {
	listOpts := cfclient.NewServiceCredentialBindingListOptions()
	listOpts.LabelSelector.EqualTo(fmt.Sprintf("%s/%s=%s", labelPrefix, labelKeyOwner, bo.owner))
	return listOpts
}

func (c *spaceClient) GetBinding(ctx context.Context, filter BindindingFilter) (*facade.Binding, error) {
  listOpts := filter.GetListOptions()
  //...
```

This way above in the Reconcile, you could just call the function like so `GetBinding(ctx, BindingFilterOwner{name: "foo"})` or `GetBinding(BindingFilterOwner{owner: "bar"}`. This is nice, since the type is checked at compile time, so we don't need the runtime check for the string. It also makes the function signature nicer and removes the need to initialize a complex string map.